### PR TITLE
JDK-8301498: Replace NULL with nullptr in cpu/x86

### PR DESCRIPTION
**Pull request by jdksjolen**
Hi, this PR changes all occurrences of NULL to nullptr for the subdirectory cpu/x86. Unfortunately the script that does the change isn't perfect, and so we
need to comb through these manually to make sure nothing has gone wrong. I also review these changes but things slip past my eyes sometimes.

Here are some typical things to look out for:

1. No changes but copyright header changed (probably because I reverted some changes but forgot the copyright).
2. Macros having their NULL changed to nullptr, these are added to the script when I find them. They should be NULL.
3. nullptr in comments and logs. We try to use lower case "null" in these cases as it reads better. An exception is made when code expressions are in a comment.

An example of this:

```c++
// This function returns null
void* ret_null();
// This function returns true if *x == nullptr
bool is_nullptr(void** x);
```

Note how `nullptr` participates in a code expression here, we really are talking about the specific value `nullptr`.

Thanks!

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8301498](https://bugs.openjdk.org/browse/JDK-8301498): Replace NULL with nullptr in cpu/x86

### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**) ⚠️ Review applies to [0c18b898](https://git.openjdk.org/jdk/pull/12326/files/0c18b898e651c2e4ab44821beafd953bd12b632d)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/12326/head:pull/12326` \
`$ git checkout pull/12326`

Update a local copy of the PR: \
`$ git checkout pull/12326` \
`$ git pull https://git.openjdk.org/jdk.git pull/12326/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 12326`

View PR using the GUI difftool: \
`$ git pr show -t 12326`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/12326.diff">https://git.openjdk.org/jdk/pull/12326.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/cpu/x86/frame_x86.inline.hpp [hunk 1]
```diff
@@ -160,19 +160,19 @@ inline frame::frame(intptr_t* sp, intptr_t* fp) {
   // Then we could use the assert below. However this assert is of somewhat dubious
   // value.
   // UPDATE: this constructor is only used by trace_method_handle_stub() now.
-  // assert(_pc != NULL, "no pc?");
+  // assert(_pc != null, "no pc?");
```
Comment (jdksjolen) [inline src/hotspot/cpu/x86/frame_x86.inline.hpp:163]: nullptr

File: src/hotspot/cpu/x86/gc/g1/g1BarrierSetAssembler_x86.cpp [hunk 2]
```diff
@@ -230,7 +230,7 @@ void G1BarrierSetAssembler::g1_write_barrier_pre(MacroAssembler* masm,
 
   // Calling the runtime using the regular call_VM_leaf mechanism generates
   // code (generated by InterpreterMacroAssember::call_VM_leaf_base)
-  // that checks that the *(ebp+frame::interpreter_frame_last_sp) == NULL.
+  // that checks that the *(ebp+frame::interpreter_frame_last_sp) == null.
```
Comment (jdksjolen) [inline src/hotspot/cpu/x86/gc/g1/g1BarrierSetAssembler_x86.cpp:233]: nullptr

File: src/hotspot/cpu/x86/gc/shenandoah/shenandoahBarrierSetAssembler_x86.cpp [hunk 3]
```diff
@@ -264,7 +264,7 @@ void ShenandoahBarrierSetAssembler::satb_write_barrier_pre(MacroAssembler* masm,
 
   // Calling the runtime using the regular call_VM_leaf mechanism generates
   // code (generated by InterpreterMacroAssember::call_VM_leaf_base)
-  // that checks that the *(ebp+frame::interpreter_frame_last_sp) == NULL.
+  // that checks that the *(ebp+frame::interpreter_frame_last_sp) == null.
```
Comment (jdksjolen) [inline src/hotspot/cpu/x86/gc/shenandoah/shenandoahBarrierSetAssembler_x86.cpp:267]: nullptr

File: src/hotspot/cpu/x86/icBuffer_x86.cpp [hunk 4]
```diff
@@ -57,7 +57,7 @@ void InlineCacheBuffer::assemble_ic_buffer_code(address code_begin, void* cached
   // because
   // (1) the value is old (i.e., doesn't matter for scavenges)
   // (2) these ICStubs are removed *before* a GC happens, so the roots disappear
-  // assert(cached_value == NULL || cached_oop->is_perm(), "must be perm oop");
+  // assert(cached_value == null || cached_oop->is_perm(), "must be perm oop");
```
Comment (jdksjolen) [inline src/hotspot/cpu/x86/icBuffer_x86.cpp:60]: nullptr

File: src/hotspot/cpu/x86/interp_masm_x86.cpp [hunk 5]
```diff
@@ -300,7 +300,7 @@ void InterpreterMacroAssembler::call_VM_base(Register oop_result,
     cmpptr(Address(rbp, frame::interpreter_frame_last_sp_offset * wordSize), NULL_WORD);
     jcc(Assembler::equal, L);
     stop("InterpreterMacroAssembler::call_VM_base:"
-         " last_sp != NULL");
+         " last_sp != null");
```
Comment (jdksjolen) [inline src/hotspot/cpu/x86/interp_masm_x86.cpp:303]: nullptr

File: src/hotspot/cpu/x86/interp_masm_x86.cpp [hunk 6]
```diff
@@ -399,7 +399,7 @@ void InterpreterMacroAssembler::check_and_handle_earlyret(Register java_thread)
 
     movptr(tmp, Address(rthread, JavaThread::jvmti_thread_state_offset()));
     testptr(tmp, tmp);
-    jcc(Assembler::zero, L); // if (thread->jvmti_thread_state() == NULL) exit;
+    jcc(Assembler::zero, L); // if (thread->jvmti_thread_state() == null) exit;
```
Comment (jdksjolen) [inline src/hotspot/cpu/x86/interp_masm_x86.cpp:402]: nullptr

File: src/hotspot/cpu/x86/interp_masm_x86.cpp [hunk 7]
```diff
@@ -1776,13 +1776,13 @@ void InterpreterMacroAssembler::record_item_in_profile_helper(Register item, Reg
 // Example state machine code for three profile rows:
 //   // main copy of decision tree, rooted at row[1]
 //   if (row[0].rec == rec) { row[0].incr(); goto done; }
-//   if (row[0].rec != NULL) {
+//   if (row[0].rec != null) {
```
Comment (jdksjolen) [inline src/hotspot/cpu/x86/interp_masm_x86.cpp:1779]: nullptr

File: src/hotspot/cpu/x86/macroAssembler_x86.cpp [hunk 8]
```diff
@@ -2863,7 +2863,7 @@ void MacroAssembler::null_check(Register reg, int offset) {
     //       testl needs to be implemented first)
   } else {
     // nothing to do, (later) access of M[reg + offset]
-    // will provoke OS NULL exception if reg = NULL
+    // will provoke OS null exception if reg = null
```
Comment (jdksjolen) [inline src/hotspot/cpu/x86/macroAssembler_x86.cpp:2866]: is not =

File: src/hotspot/cpu/x86/macroAssembler_x86.cpp [hunk 9]
```diff
@@ -4262,7 +4262,7 @@ void MacroAssembler::lookup_interface_method(Register recv_klass,
     lea(recv_klass, Address(recv_klass, itable_index, Address::times_ptr, itentry_off));
   }
 
-  // for (scan = klass->itable(); scan->interface() != NULL; scan += scan_step) {
+  // for (scan = klass->itable(); scan->interface() != null; scan += scan_step) {
```
Comment (jdksjolen) [inline src/hotspot/cpu/x86/macroAssembler_x86.cpp:4265]: nullptr

File: src/hotspot/cpu/x86/stubGenerator_x86_32.cpp [hunk 10]
```diff
@@ -1544,13 +1544,13 @@ class StubGenerator: public StubCodeGenerator {
...
 #define LOCAL_JCC(assembler_con, label_ptr)                             \
-    if (label_ptr != NULL)  __ jcc(assembler_con, *(label_ptr));        \
+    if (label_ptr != nullptr)  __ jcc(assembler_con, *(label_ptr));        \
     else                    __ jcc(assembler_con, L_fallthrough) /*omit semi*/
 
     // The following is a strange variation of the fast path which requires
     // one less register, because needed values are on the argument stack.
     // __ check_klass_subtype_fast_path(sub_klass, *super_klass*, temp,
-    //                                  L_success, L_failure, NULL);
+    //                                  L_success, L_failure, null);
```
Comment (jdksjolen) [inline src/hotspot/cpu/x86/stubGenerator_x86_32.cpp:1553]: nullptr

File: src/hotspot/cpu/x86/stubGenerator_x86_32.cpp [hunk 11]
```diff
@@ -1921,7 +1921,7 @@ class StubGenerator: public StubCodeGenerator {
     const Register dst_pos = rdi;
     const Register length  = rcx;       // transfer count
 
-    //  if (src == NULL) return -1;
+    //  if (src == null) return -1;
```
Comment (jdksjolen) [inline src/hotspot/cpu/x86/stubGenerator_x86_32.cpp:1924]: nullptr

File: src/hotspot/cpu/x86/stubGenerator_x86_32.cpp [hunk 12]
```diff
@@ -1931,7 +1931,7 @@ class StubGenerator: public StubCodeGenerator {
     __ testl(src_pos, src_pos);
     __ jccb(Assembler::negative, L_failed_0);
 
-    //  if (dst == NULL) return -1;
+    //  if (dst == null) return -1;
```
Comment (jdksjolen) [inline src/hotspot/cpu/x86/stubGenerator_x86_32.cpp:1934]: nullptr

File: src/hotspot/cpu/x86/stubGenerator_x86_32.cpp [hunk 13]
```diff
@@ -1946,18 +1946,18 @@ class StubGenerator: public StubCodeGenerator {
...
-    //  if (src->klass() == NULL) return -1;
+    //  if (src->klass() == null) return -1;
     Address src_klass_addr(src, oopDesc::klass_offset_in_bytes());
     Address dst_klass_addr(dst, oopDesc::klass_offset_in_bytes());
     const Register rcx_src_klass = rcx;    // array klass
     __ movptr(rcx_src_klass, Address(src, oopDesc::klass_offset_in_bytes()));
 
 #ifdef ASSERT
-    //  assert(src->klass() != NULL);
+    //  assert(src->klass() != null);
```
Comment (jdksjolen) [inline src/hotspot/cpu/x86/stubGenerator_x86_32.cpp:1956]: nullptr

File: src/hotspot/cpu/x86/stubGenerator_x86_64_arraycopy.cpp [hunk 14]
```diff
@@ -2289,13 +2289,13 @@ address StubGenerator::generate_generic_copy(const char *name,
...
   // (4) length  must not be negative.
-  // (5) src klass and dst klass should be the same and not NULL.
+  // (5) src klass and dst klass should be the same and not null.
   // (6) src and dst should be arrays.
   // (7) src_pos + length must not exceed length of src.
   // (8) dst_pos + length must not exceed length of dst.
   //
 
-  //  if (src == NULL) return -1;
+  //  if (src == null) return -1;
```
Comment (jdksjolen) [inline src/hotspot/cpu/x86/stubGenerator_x86_64_arraycopy.cpp:2298]: nullptr

File: src/hotspot/cpu/x86/stubGenerator_x86_64_arraycopy.cpp [hunk 15]
```diff
@@ -2304,7 +2304,7 @@ address StubGenerator::generate_generic_copy(const char *name,
   __ testl(src_pos, src_pos); // src_pos (32-bits)
   __ jccb(Assembler::negative, L_failed_0);
 
-  //  if (dst == NULL) return -1;
+  //  if (dst == null) return -1;
```
Comment (jdksjolen) [inline src/hotspot/cpu/x86/stubGenerator_x86_64_arraycopy.cpp:2307]: nullptr

File: src/hotspot/cpu/x86/stubGenerator_x86_64_arraycopy.cpp [hunk 16]
```diff
@@ -2332,12 +2332,12 @@ address StubGenerator::generate_generic_copy(const char *name,
 
   __ load_klass(r10_src_klass, src, rklass_tmp);
 #ifdef ASSERT
-  //  assert(src->klass() != NULL);
+  //  assert(src->klass() != null);
```
Comment (jdksjolen) [inline src/hotspot/cpu/x86/stubGenerator_x86_64_arraycopy.cpp:2335]: nullptr

**Review by jdksjolen — COMMENTED** [review]
Some things to fix

**Comment by jdksjolen** [comment]
Builds and sent out for testing.

**Comment by jdksjolen** [comment]
Passes tier1.

File: src/hotspot/cpu/x86/interp_masm_x86.cpp [hunk 17]
```diff
@@ -300,7 +300,7 @@ void InterpreterMacroAssembler::call_VM_base(Register oop_result,
     cmpptr(Address(rbp, frame::interpreter_frame_last_sp_offset * wordSize), NULL_WORD);
     jcc(Assembler::equal, L);
     stop("InterpreterMacroAssembler::call_VM_base:"
-         " last_sp != NULL");
+         " last_sp != nullptr");
```
Comment (vnkozlov) [inline src/hotspot/cpu/x86/interp_masm_x86.cpp:303]: `null`

Comment (dholmes-ora) [reply src/hotspot/cpu/x86/interp_masm_x86.cpp:303]: These are textual code fragments so I think `nullptr` is more appropriate - as per changes made to cpu/arm code.

File: src/hotspot/cpu/x86/interp_masm_x86.cpp [hunk 18]
```diff
@@ -268,7 +268,7 @@ void InterpreterMacroAssembler::call_VM_leaf_base(address entry_point,
     cmpptr(Address(rbp, frame::interpreter_frame_last_sp_offset * wordSize), NULL_WORD);
     jcc(Assembler::equal, L);
     stop("InterpreterMacroAssembler::call_VM_leaf_base:"
-         " last_sp != NULL");
+         " last_sp != null");
```
Comment (dholmes-ora) [inline src/hotspot/cpu/x86/interp_masm_x86.cpp:271]: nullptr

File: src/hotspot/cpu/x86/macroAssembler_x86.cpp [hunk 19]
```diff
@@ -2854,7 +2854,7 @@ void MacroAssembler::mulss(XMMRegister dst, AddressLiteral src, Register rscratc
 
 void MacroAssembler::null_check(Register reg, int offset) {
   if (needs_explicit_null_check(offset)) {
-    // provoke OS NULL exception if reg = NULL by
+    // provoke OS null exception if reg = null by
```
Comment (dholmes-ora) [inline src/hotspot/cpu/x86/macroAssembler_x86.cpp:2857]: suggestion "reg is null"

File: src/hotspot/cpu/x86/macroAssembler_x86.cpp (hunk 8 above)
Comment (dholmes-ora) [reply src/hotspot/cpu/x86/macroAssembler_x86.cpp:2866]: suggestion: "reg is null"

File: src/hotspot/cpu/x86/macroAssembler_x86.hpp [hunk 20]
```diff
@@ -91,9 +91,9 @@ class MacroAssembler: public Assembler {
   Address as_Address(AddressLiteral adr);
   Address as_Address(ArrayAddress adr, Register rscratch);
 
-  // Support for NULL-checks
+  // Support for null-checks
   //
-  // Generates code that causes a NULL OS exception if the content of reg is NULL.
+  // Generates code that causes a null OS exception if the content of reg is null.
```
Comment (dholmes-ora) [inline src/hotspot/cpu/x86/macroAssembler_x86.hpp:96]: Note use of "reg is null" here :)

File: src/hotspot/cpu/x86/sharedRuntime_x86_64.cpp [hunk 21]
```diff
@@ -1762,10 +1762,10 @@ nmethod* SharedRuntime::generate_native_wrapper(MacroAssembler* masm,
                                        stack_slots / VMRegImpl::slots_per_word,
                                        in_ByteSize(-1),
                                        in_ByteSize(-1),
-                                       (OopMapSet*)NULL);
+                                       (OopMapSet*)nullptr);
```
Comment (dholmes-ora) [inline src/hotspot/cpu/x86/sharedRuntime_x86_64.cpp:1765]: Unnecessary cast?

File: src/hotspot/cpu/x86/templateTable_x86.cpp [hunk 22]
```diff
@@ -4184,8 +4184,8 @@ void TemplateTable::instanceof() {
     __ bind(is_null);   // same as 'done'
   }
   __ bind(done);
-  // rax = 0: obj == NULL or  obj is not an instanceof the specified klass
-  // rax = 1: obj != NULL and obj is     an instanceof the specified klass
+  // rax = 0: obj == null or  obj is not an instanceof the specified klass
+  // rax = 1: obj != null and obj is     an instanceof the specified klass
```
Comment (dholmes-ora) [inline src/hotspot/cpu/x86/templateTable_x86.cpp:4188]: nullptr

**Comment by jdksjolen** [comment]
I added the fixes, thanks! Last commit passed tier1.

File: src/hotspot/cpu/x86/interp_masm_x86.cpp (hunk 17 above)
Comment (vnkozlov) [reply src/hotspot/cpu/x86/interp_masm_x86.cpp:303]: Right.

**Review by dholmes-ora — APPROVED** [review]
LGTM! Thanks

File: src/hotspot/cpu/x86/interp_masm_x86.cpp (hunk 18 above)
Comment (vnkozlov) [reply src/hotspot/cpu/x86/interp_masm_x86.cpp:271]: Missed change to nullptr as David suggested.

File: src/hotspot/cpu/x86/interp_masm_x86.cpp (hunk 6 above)
Comment (vnkozlov) [reply src/hotspot/cpu/x86/interp_masm_x86.cpp:402]: Missed change to nullptr as you suggested.

File: src/hotspot/cpu/x86/interp_masm_x86.cpp (hunk 7 above)
Comment (vnkozlov) [reply src/hotspot/cpu/x86/interp_masm_x86.cpp:1779]: Missed change to nullptr as you suggested.

File: src/hotspot/cpu/x86/interp_masm_x86.cpp [hunk 23]
```diff
@@ -1776,13 +1776,13 @@ void InterpreterMacroAssembler::record_item_in_profile_helper(Register item, Reg
...
-//   if (row[0].rec != NULL) {
+//   if (row[0].rec != null) {
 //     // inner copy of decision tree, rooted at row[1]
 //     if (row[1].rec == rec) { row[1].incr(); goto done; }
-//     if (row[1].rec != NULL) {
+//     if (row[1].rec != null) {
 //       // degenerate decision tree, rooted at row[2]
 //       if (row[2].rec == rec) { row[2].incr(); goto done; }
-//       if (row[2].rec != NULL) { count.incr(); goto done; } // overflow
+//       if (row[2].rec != null) { count.incr(); goto done; } // overflow
```
Comment (vnkozlov) [inline src/hotspot/cpu/x86/interp_masm_x86.cpp:1785]: Missed change to nullptr.

File: src/hotspot/cpu/x86/macroAssembler_x86.cpp (hunk 9 above)
Comment (vnkozlov) [reply src/hotspot/cpu/x86/macroAssembler_x86.cpp:4265]: Missed change to nullptr as you suggested.

**Review by vnkozlov — CHANGES_REQUESTED** [review]
In addition to few missed placed I noticed that you keep `mov_metadata(rbx, (Metadata*)nullptr);` casting.

Is `Metadata*` something special we need to cast it?

And more general question - do we need to keep some casting for `nullptr`?
We kept them in previous PRs which are already pushed #12029 

Should we update our "spec" for `nullptr` to remove all casting or keep some? What is condition to keep?

**Comment by jdksjolen** [comment]
Hi Vladimir!

I applied your changes, thanks for the review.

**Review by vnkozlov — APPROVED** [review]
Looks good.

**Review by dholmes-ora — APPROVED** [review]
Still good. Thanks.

**Comment by jdksjolen** [comment]
Thanks. All of tier1 and tier2 is passing.

/integrate

**Pull request closed without merging**